### PR TITLE
KAFKA-14869: Ignore unknown record types for coordinators (KIP-915, Part-1)

### PR DESCRIPTION
**Pull request by jeffkbkim**
KIP-915: Part 1 of 2

Coordinators should Ignore unknown record types

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jeffkbkim** [comment]
will re-trigger build once https://github.com/apache/kafka/pull/13512 is merged. all test failures are related to streams

File: clients/src/main/java/org/apache/kafka/common/protocol/MessageUtil.java [hunk 1]
```diff
@@ -231,4 +231,38 @@ public static byte[] toVersionPrefixedBytes(final short version, final Message m
                 buffer.limit() == buffer.array().length) return buffer.array();
         else return Utils.toArray(buffer);
     }
+
+    // Should only be used for testing
+    public static byte[] messageWithUnknownVersion() {
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/common/protocol/MessageUtil.java]: It is a bit awkward to have this in the `src` directory. I am not really sure to understand why we need it as well. I suppose that we could use any other `Message` like record from `test`, isn't it?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 2]
```diff
@@ -1145,17 +1150,22 @@ object GroupMetadataManager {
    * @param buffer input byte-buffer
    * @return an OffsetKey or GroupMetadataKey object from the message
    */
-  def readMessageKey(buffer: ByteBuffer): BaseKey = {
+  def readMessageKey(buffer: ByteBuffer): Option[BaseKey] = {
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: Have you considered introducing an `UnknownKey` which extends `BaseKey`? The `UnknownKey` could carry on the version and even the bytes. It seems that the change would be less disruptive like this.

**Review by dajac — COMMENTED** [review]
@jeffkbkim Thanks for the patch. I made a first pass and I left a few suggestions.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 3]
```diff
@@ -683,8 +685,10 @@ class GroupMetadataManager(brokerId: Int,
                       removedGroups.add(groupId)
                     }
 
-                  case unknownKey =>
-                    throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
+                  case _: UnknownKey => // do nothing
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: Should we put the warning here instead of having it in `readMessageKey`?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 4]
```diff
@@ -683,8 +685,10 @@ class GroupMetadataManager(brokerId: Int,
...
+
+                  case unexpectedKey =>
+                    throw new IllegalStateException(s"Unexpected message key $unexpectedKey while loading offsets and group metadata")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: Is this one still required?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 5]
```diff
@@ -1155,7 +1159,12 @@ object GroupMetadataManager {
       // version 2 refers to group metadata
       val key = new GroupMetadataKeyData(new ByteBufferAccessor(buffer), version)
       GroupMetadataKey(version, key.group)
-    } else throw new IllegalStateException(s"Unknown group metadata message version: $version")
+    } else {
+      // Unknown versions may exist when a downgraded coordinator is reading records from the log.
+      warn(s"Found unknown message key version: $version." +
+        s" The downgraded coordinator will ignore this key and corresponding value.")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: `downgraded coordinator` reads a bit weird here. How about: `Unexpected message key with version ($version) while loading offsets and group metadata. Ignoring it.`?

I wonder if we should put a sentence like `It could be a left over from an aborted upgrade.`. What do you think?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 6]
```diff
@@ -1273,9 +1282,10 @@ object GroupMetadataManager {
       throw new KafkaException("Failed to decode message using offset topic decoder (message had a missing key)")
     } else {
       GroupMetadataManager.readMessageKey(record.key) match {
-        case offsetKey: OffsetKey => parseOffsets(offsetKey, record.value)
-        case groupMetadataKey: GroupMetadataKey => parseGroupMetadata(groupMetadataKey, record.value)
-        case _ => throw new KafkaException("Failed to decode message using offset topic decoder (message had an invalid key)")
+          case offsetKey: OffsetKey => parseOffsets(offsetKey, record.value)
+          case groupMetadataKey: GroupMetadataKey => parseGroupMetadata(groupMetadataKey, record.value)
+          case _: UnknownKey => (Some("<UNKNOWN>"), Some("<UNKNOWN>"))
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: For the key, could we say `Unknown(version=$version`? For the value, could we just return `None`?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 7]
```diff
@@ -1368,3 +1378,8 @@ case class GroupMetadataKey(version: Short, key: String) extends BaseKey {
   override def toString: String = key
 }
 
+case class UnknownKey(version: Short, key: String = null) extends BaseKey {
+
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: I would remove this empty line.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala [hunk 8]
```diff
@@ -640,8 +640,13 @@ class GroupMetadataManagerTest {
     val offsetCommitRecords = createCommittedOffsetRecords(committedOffsets)
     val memberId = "98098230493"
     val groupMetadataRecord = buildStableGroupRecordWithMember(generation, protocolType, protocol, memberId)
+
+    // Should ignore unknown record
+    val unknownMessage = MessageUtil.messageWithUnknownVersion()
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala]: Does this method still exist?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala [hunk 9]
```diff
@@ -2762,4 +2767,12 @@ class GroupMetadataManagerTest {
     assertTrue(partitionLoadTime("partition-load-time-max") >= diff)
     assertTrue(partitionLoadTime("partition-load-time-avg") >= diff)
   }
+
+  @Test
+  def testIgnoreUnknownMessageKeyVersion(): Unit = {
+    val record = new org.apache.kafka.coordinator.group.generated.GroupMetadataKey()
+    val unknownRecord = MessageUtil.toVersionPrefixedBytes(Short.MaxValue, record)
+    GroupMetadataManager.readMessageKey(ByteBuffer.wrap(unknownRecord))
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala]: nit: Could we assert the returned value?

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionLogTest.scala [hunk 10]
```diff
@@ -135,4 +138,11 @@ class TransactionLogTest {
     assertEquals(Some("<DELETE>"), valueStringOpt)
   }
 
+  @Test
+  def testReadUnknownMessageKeyVersion(): Unit = {
+    val record = new TransactionLogKey()
+    val unknownRecord = MessageUtil.toVersionPrefixedBytes(Short.MaxValue, record)
+    TransactionLog.readTxnRecordKey(ByteBuffer.wrap(unknownRecord))
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/transaction/TransactionLogTest.scala]: ditto.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala [hunk 11]
```diff
@@ -148,17 +152,23 @@ object TransactionLog {
...
+          val producerIdMetadata = if (value == null)
+            None
+          else
+            readTxnRecordValue(transactionalId, ByteBuffer.wrap(value))
+          output.write(transactionalId.getBytes(StandardCharsets.UTF_8))
+          output.write("::".getBytes(StandardCharsets.UTF_8))
+          output.write(producerIdMetadata.getOrElse("NULL").toString.getBytes(StandardCharsets.UTF_8))
+          output.write("\n".getBytes(StandardCharsets.UTF_8))
+
+        case _: UnknownKey => // Only print if this message is a transaction record
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: nit: I would also add the logging here.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala [hunk 12]
```diff
@@ -148,17 +152,23 @@ object TransactionLog {
...
+
+        case unexpectedKey =>
+          throw new IllegalStateException(s"Found unexpected key $unexpectedKey while reading transaction log.")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: Same question about this one.

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala [hunk 13]
```diff
@@ -758,7 +758,7 @@ class TransactionStateManagerTest {
     appendedRecords.values.foreach { batches =>
       batches.foreach { records =>
         records.records.forEach { record =>
-          val transactionalId = TransactionLog.readTxnRecordKey(record.key).transactionalId
+          val transactionalId = TransactionLog.readTxnRecordKey(record.key).get.transactionalId
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala]: I am not sure to understand why we need this `get` here.

**Review by dajac — COMMENTED** [review]
@jeffkbkim Thanks for the update. A few more comments.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala (hunk 11 above)
Comment (Hangleton) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: Should we log or print it to notify the user of the presence of the key?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala [hunk 14]
```diff
@@ -167,25 +177,43 @@ object TransactionLog {
...
+trait BaseKey{
+  def version: Short
+  def transactionalId: String
+}
+
+case class TxnKey(version: Short, transactionalId: String) extends BaseKey {
   override def toString: String = transactionalId
 }
+
+case class UnknownKey(version: Short, transactionalId: String = null) extends BaseKey {
```
Comment (Hangleton) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: Is the `transactionalId` going to be populated?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 15]
```diff
@@ -1368,3 +1378,8 @@ case class GroupMetadataKey(version: Short, key: String) extends BaseKey {
   override def toString: String = key
 }
 
+case class UnknownKey(version: Short, key: String = null) extends BaseKey {
```
Comment (Hangleton) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: Is the `key` going to be populated?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala (hunk 5 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: makes sense, will use Unknown instead of Unexpected

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala (hunk 15 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: thanks. removed the field

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala (hunk 4 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: we need this: `match may not be exhaustive.`

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala (hunk 12 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: we need this: `match may not be exhaustive.`

Comment (dajac) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: Does it help if we use a « sealed trait »?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 16]
```diff
@@ -576,7 +576,10 @@ class GroupMetadataManager(brokerId: Int,
     }
   }
 
-  private def doLoadGroupsAndOffsets(topicPartition: TopicPartition, onGroupLoaded: GroupMetadata => Unit): Unit = {
+  // Visible for testing
+  private[group] def doLoadGroupsAndOffsets(topicPartition: TopicPartition,
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: It seems that we could revert this change now.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala (hunk 4 above)
Comment (dajac) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: I think that we could use a `sealed trait` to fix this. Does it work?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 17]
```diff
@@ -1273,9 +1283,10 @@ object GroupMetadataManager {
       throw new KafkaException("Failed to decode message using offset topic decoder (message had a missing key)")
     } else {
       GroupMetadataManager.readMessageKey(record.key) match {
-        case offsetKey: OffsetKey => parseOffsets(offsetKey, record.value)
-        case groupMetadataKey: GroupMetadataKey => parseGroupMetadata(groupMetadataKey, record.value)
-        case _ => throw new KafkaException("Failed to decode message using offset topic decoder (message had an invalid key)")
+          case offsetKey: OffsetKey => parseOffsets(offsetKey, record.value)
+          case groupMetadataKey: GroupMetadataKey => parseGroupMetadata(groupMetadataKey, record.value)
+          case unknownKey: UnknownKey => (Some(s"UNKNOWN(version=${unknownKey.version})"), None)
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: In order to follow the format of the offset and group keys, I wonder if we should use `unknown::version=$version`. What do you think?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala [hunk 18]
```diff
@@ -148,17 +149,23 @@ object TransactionLog {
...
+            readTxnRecordValue(transactionalId, ByteBuffer.wrap(value))
+          output.write(transactionalId.getBytes(StandardCharsets.UTF_8))
+          output.write("::".getBytes(StandardCharsets.UTF_8))
+          output.write(producerIdMetadata.getOrElse("NULL").toString.getBytes(StandardCharsets.UTF_8))
+          output.write("\n".getBytes(StandardCharsets.UTF_8))
+
+        case _: UnknownKey => // Only print if this message is a transaction record
+
+        case unexpectedKey =>
+          throw new IllegalStateException(s"Found unexpected key $unexpectedKey while reading transaction log.")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: `writeTo` is actually used to write records to the log. It does not make sense to swallow unknown records here. Therefore, we should throw an `IllegalStateException` for unknown records as well. Moreover, the message should be change as `while reading transaction log` is not correct.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala [hunk 19]
```diff
@@ -167,25 +174,44 @@ object TransactionLog {
...
+            s"state=${txnMetadata.state}," +
+            s"partitions=${txnMetadata.topicPartitions.mkString("[", ",", "]")}," +
+            s"txnLastUpdateTimestamp=${txnMetadata.txnLastUpdateTimestamp}," +
+            s"txnTimeoutMs=${txnMetadata.txnTimeoutMs}"
+        }
+
+        (Some(keyString), Some(valueString))
+
+      case _: UnknownKey =>
+        (Some("<UNKNOWN>"), Some("<UNKNOWN>"))
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: Let's use the same format than we used for the groups.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala [hunk 20]
```diff
@@ -167,25 +174,44 @@ object TransactionLog {
...
 
-    (Some(keyString), Some(valueString))
+      case unexpectedKey =>
+        throw new IllegalStateException(s"Found unexpected key $unexpectedKey while formatting transaction log.")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: I think that we could remove this one if we use a `sealed trait`.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 21]
```diff
@@ -467,16 +467,26 @@ class TransactionStateManager(brokerId: Int,
...
+                    }
+                    currOffset = batch.nextOffset
+
+                  case unknownKey: UnknownKey =>
+                    warn(s"Unknown message key with version ${unknownKey.version}" +
+                      s" while loading transaction state. Ignoring it. " +
+                      s"It could be a left over from an aborted upgrade.")
+
+                  case unexpectedKey =>
+                    throw new IllegalStateException(s"Found unexpected key $unexpectedKey while reading transaction log.")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: I think that we could remove this one if we use a `sealed trait`.

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala [hunk 22]
```diff
@@ -1085,4 +1087,41 @@ class TransactionStateManagerTest {
     assertTrue(partitionLoadTime("partition-load-time-max") >= 0)
     assertTrue(partitionLoadTime( "partition-load-time-avg") >= 0)
   }
+
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/transaction/TransactionStateManagerTest.scala:1090]: nit: There is an extra empty line.

**Review by dajac — COMMENTED** [review]
@jeffkbkim Thanks for the update. I left a few more comments.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala (hunk 4 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: that works. thanks!

File: core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala (hunk 18 above)
Comment (jeffkbkim) [reply core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala]: this is used to write to output stream in console consumer. shouldn’t we be printing unknown::version=$version here as well? i agree the log wording needs to be changed. i have replaced the exception with a print for now.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 23]
```diff
@@ -683,8 +682,11 @@ class GroupMetadataManager(brokerId: Int,
                       removedGroups.add(groupId)
                     }
 
-                  case unknownKey =>
-                    throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
+                  case unknownKey: UnknownKey =>
+                    // Unknown versions may exist when a downgraded coordinator is reading records from the log.
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: We could remove this comment as the warning contains everything.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 24]
```diff
@@ -683,8 +682,11 @@ class GroupMetadataManager(brokerId: Int,
...
+                    warn(s"Unknown message key with version ${unknownKey.version}" +
+                      s" while loading offsets and group metadata. Ignoring it. " +
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: I wonder if we could add `while loading offsets and group metadata from $topicPartition`. Having the partition may be useful as well.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala [hunk 25]
```diff
@@ -683,8 +682,11 @@ class GroupMetadataManager(brokerId: Int,
...
+                      s"It could be a left over from an aborted upgrade.")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala]: nit: We can remove the `s` as there is no interpolation for this one.

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 26]
```diff
@@ -467,16 +467,23 @@ class TransactionStateManager(brokerId: Int,
...
+                      case None =>
+                        loadedTransactions.remove(transactionalId)
+                      case Some(txnMetadata) =>
+                        loadedTransactions.put(transactionalId, txnMetadata)
+                    }
+                    currOffset = batch.nextOffset
+
+                  case unknownKey: UnknownKey =>
+                    warn(s"Unknown message key with version ${unknownKey.version}" +
+                      s" while loading transaction state. Ignoring it. " +
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: nit: Could we add the partition here as well?

File: core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala [hunk 27]
```diff
@@ -467,16 +467,23 @@ class TransactionStateManager(brokerId: Int,
...
+                      s"It could be a left over from an aborted upgrade.")
```
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala]: nit: `s` can be removed here as well.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala [hunk 28]
```diff
@@ -640,8 +640,14 @@ class GroupMetadataManagerTest {
     val offsetCommitRecords = createCommittedOffsetRecords(committedOffsets)
     val memberId = "98098230493"
     val groupMetadataRecord = buildStableGroupRecordWithMember(generation, protocolType, protocol, memberId)
+
+    // Should ignore unknown record
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala]: I was thinking about this one. I wonder if it would not be better to pull this into a separate unit test instead of hijacking existing ones. Is it possible?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala [hunk 29]
```diff
@@ -2762,4 +2768,12 @@ class GroupMetadataManagerTest {
     assertTrue(partitionLoadTime("partition-load-time-max") >= diff)
     assertTrue(partitionLoadTime("partition-load-time-avg") >= diff)
   }
+
+  @Test
+  def testIgnoreUnknownMessageKeyVersion(): Unit = {
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala]: nit: `testReadMessageKeyCanReadUnknownMessage`?

File: core/src/test/scala/unit/kafka/coordinator/transaction/TransactionLogTest.scala [hunk 30]
```diff
@@ -135,4 +138,11 @@ class TransactionLogTest {
     assertEquals(Some("<DELETE>"), valueStringOpt)
   }
 
+  @Test
+  def testReadUnknownMessageKeyVersion(): Unit = {
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/transaction/TransactionLogTest.scala]: nit: `testReadTxnRecordKeyCanReadUnknownMessage`?

**Review by dajac — COMMENTED** [review]
@jeffkbkim Thanks for the update. I left a few nits. We should be good to go once addressed.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala [hunk 31]
```diff
@@ -640,8 +640,14 @@ class GroupMetadataManagerTest {
...
     val groupMetadataRecord = buildStableGroupRecordWithMember(generation, protocolType, protocol, memberId)
+
+    // Should ignore unknown record
+    val unknownKey = new org.apache.kafka.coordinator.group.generated.GroupMetadataKey()
+    val unknownMessage = MessageUtil.toVersionPrefixedBytes(Short.MaxValue, unknownKey)
+    val unknownRecord = new SimpleRecord(unknownMessage, unknownMessage)
+
     val records = MemoryRecords.withRecords(startOffset, CompressionType.NONE,
-      (offsetCommitRecords ++ Seq(groupMetadataRecord)).toArray: _*)
+      (offsetCommitRecords ++ Seq(unknownRecord) ++ Seq(groupMetadataRecord)).toArray: _*)
```
Comment (Hangleton) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala]: Nit: maybe we could have a test with a few more unknown records in the list with different version.

**Review by dajac — APPROVED** [review]
LGTM, thanks. I will merge when the build completes.

**Comment by dajac** [comment]
Merged to trunk, 3.5 and 3.4. We need to open PRs for older branches.
